### PR TITLE
add Weaviate memory adapter

### PR DESCRIPTION
**Pull request by zainhas**
### Why this PR
We want to setup Weaviate as a remote vector db provider for llama-stack.

### What is in the PR

- Add in Weaviate memory adapter to support using Weaviate cloud cluster as a vector db
- Uses cluster url and api keys to connect to remote vector db cluster

### CODE REVIEW

File: llama_stack/apis/memory/client.py [hunk 1]
```diff
@@ -42,7 +42,11 @@ async def get_memory_bank(self, bank_id: str) -> Optional[MemoryBank]:
                 params={
                     "bank_id": bank_id,
                 },
-                headers={"Content-Type": "application/json"},
+                headers={
```
Comment (ashwinb) [inline llama_stack/apis/memory/client.py:45]: can we revert these changes from here, since this is a pretty generic client -- we will figure out how to get a bunch of these examples situated.

File: llama_stack/providers/adapters/memory/weaviate/__init__.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from llama_stack.distribution.datatypes import RemoteProviderConfig
+
+async def get_adapter_impl(config: RemoteProviderConfig, _deps):
+    from .weaviate import WeaviateMemoryAdapter
+
+    impl = WeaviateMemoryAdapter(config.url, config.username, config.password)
```
Comment (ashwinb) [inline llama_stack/providers/adapters/memory/weaviate/__init__.py:6]: the config does not define url / username / password but instead only defines a collection?

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+class WeaviateMemoryAdapter(Memory):
+    def __init__(self, config: WeaviateConfig) -> None:
+        self.config = config
+        self.client = None
+        self.cache = {}
+
+    async def initialize(self) -> None:
+        try:
+
+            request_provider_data = get_request_provider_data()
```
Comment (ashwinb) [inline llama_stack/providers/adapters/memory/weaviate/weaviate.py:81]: provider data isn't present at this point -- it is only provided at the time of the request. you should initialize the client on every client call and if we need a cache of clients then, we'd need to build that.

**Review by ashwinb — COMMENTED** [review]
A few important comments inline. But overall:

- let's figure out a way to test this end to end
- also, could you please install pre-commit and lint / format the code

File: llama_stack/apis/memory/client.py (hunk 1 above)
Comment (zainhas) [reply llama_stack/apis/memory/client.py:45]: done

File: llama_stack/providers/adapters/memory/weaviate/__init__.py (hunk 2 above)
Comment (zainhas) [reply llama_stack/providers/adapters/memory/weaviate/__init__.py:6]: the cluster url and api_key for weaviate will come in via the WeaviateRequestProviderData class, in this case should I just get rid of the WeaviateConfig class?

Comment (zainhas) [reply llama_stack/providers/adapters/memory/weaviate/__init__.py:6]: done

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py (hunk 3 above)
Comment (zainhas) [reply llama_stack/providers/adapters/memory/weaviate/weaviate.py:81]: modified the method to return an initialized client. following this I initialize the client on every client call using the method.

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    def __init__(self, client: weaviate.Client, collection: str):
+        self.client = client
+        self.collection = collection
+
+    async def add_chunks(self, chunks: List[Chunk], embeddings: NDArray):
+        assert len(chunks) == len(embeddings), f"Chunk length {len(chunks)} does not match embedding length {len(embeddings)}"
+
+        data_objects = []
+        for i, chunk in enumerate(chunks):
+            print(f"Adding chunk #{i} tokens={chunk.token_count}")
```
Comment (codefromthecrypt) [inline llama_stack/providers/adapters/memory/weaviate/weaviate.py:29]: curious: are we ok with print logging?

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        self.client = None
+        self.cache = {}
+
+    def _get_client(self) -> weaviate.Client:
+            request_provider_data = get_request_provider_data()
+            
+            if request_provider_data is not None:
+                assert isinstance(request_provider_data, WeaviateRequestProviderData)
+
+                print(f"WEAVIATE API KEY: {request_provider_data.weaviate_api_key}")
```
Comment (codefromthecrypt) [inline llama_stack/providers/adapters/memory/weaviate/weaviate.py:84]: do we really want to print the api key into the console?

**Review by codefromthecrypt — COMMENTED** [review]
also curious about tests, possibly using vcr might work so that folks don't need a real endpoint?

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py (hunk 4 above)
Comment (zainhas) [reply llama_stack/providers/adapters/memory/weaviate/weaviate.py]: updated to remove this

File: llama_stack/providers/adapters/memory/weaviate/weaviate.py (hunk 5 above)
Comment (zainhas) [reply llama_stack/providers/adapters/memory/weaviate/weaviate.py]: removed

**Review by ashwinb — APPROVED** [review]
I am merging this in and then will fix whatever doesn't quite work as part of #201
